// @inworld/nodejs-sdk/build/src/entities/emotion-behavior.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmotionBehavior = void 0;
const packets_pb_1 = require("@proto/packets_pb");
class EmotionBehavior {
    constructor(behavior) {
        this.behavior = behavior;
    }
    isNeutral() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.NEUTRAL;
    }
    isDisgust() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.DISGUST;
    }
    isContempt() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.CONTEMPT;
    }
    isBelligerence() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.BELLIGERENCE;
    }
    isDomineering() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.DOMINEERING;
    }
    isCriticism() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.CRITICISM;
    }
    isAnger() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.ANGER;
    }
    isTension() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.TENSION;
    }
    isTenseHumor() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.TENSE_HUMOR;
    }
    isDefensiveness() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.DEFENSIVENESS;
    }
    isWhining() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.WHINING;
    }
    isSadness() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.SADNESS;
    }
    isStonewalling() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.STONEWALLING;
    }
    isInterest() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.INTEREST;
    }
    isValidation() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.VALIDATION;
    }
    isAffection() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.AFFECTION;
    }
    isHumor() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.HUMOR;
    }
    isSurprise() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.SURPRISE;
    }
    isJoy() {
        return this.behavior === packets_pb_1.EmotionEvent.SpaffCode.JOY;
    }
}
exports.EmotionBehavior = EmotionBehavior;
