// @inworld/nodejs-sdk/build/src/entities/inworld_packet.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InworldPacket = exports.InworlControlType = exports.InworldPacketType = void 0;
var InworldPacketType;
(function (InworldPacketType) {
    InworldPacketType[InworldPacketType["UNKNOWN"] = 0] = "UNKNOWN";
    InworldPacketType[InworldPacketType["TEXT"] = 1] = "TEXT";
    InworldPacketType[InworldPacketType["AUDIO"] = 2] = "AUDIO";
    InworldPacketType[InworldPacketType["CUSTOM"] = 3] = "CUSTOM";
    InworldPacketType[InworldPacketType["EMOTION"] = 4] = "EMOTION";
    InworldPacketType[InworldPacketType["CONTROL"] = 5] = "CONTROL";
})(InworldPacketType = exports.InworldPacketType || (exports.InworldPacketType = {}));
var InworlControlType;
(function (InworlControlType) {
    InworlControlType[InworlControlType["UNKNOWN"] = 0] = "UNKNOWN";
    InworlControlType[InworlControlType["INTERACTION_END"] = 3] = "INTERACTION_END";
})(InworlControlType = exports.InworlControlType || (exports.InworlControlType = {}));
class InworldPacket {
    constructor(props) {
        this.packetId = props.packetId;
        this.routing = props.routing;
        this.date = props.date;
        this.type = props.type;
        if (this.isText()) {
            this.text = props.text;
        }
        if (this.isAudio()) {
            this.audio = props.audio;
        }
        if (this.isControl()) {
            this.control = props.control;
        }
        if (this.isEmotion()) {
            this.emotions = props.emotions;
        }
        if (this.isCustom()) {
            this.custom = props.custom;
        }
    }
    isText() {
        return this.type === InworldPacketType.TEXT;
    }
    isAudio() {
        return this.type === InworldPacketType.AUDIO;
    }
    isControl() {
        return this.type === InworldPacketType.CONTROL;
    }
    isCustom() {
        return this.type === InworldPacketType.CUSTOM;
    }
    isEmotion() {
        return this.type === InworldPacketType.EMOTION;
    }
    isInteractionEnd() {
        return (this.isControl() &&
            this.control.type === InworlControlType.INTERACTION_END);
    }
}
exports.InworldPacket = InworldPacket;
